Alternate version of Appendix with nesting example (#286)
ietf-wg-spice/draft-ietf-spice-sd-cwt: draft-ietf-spice-sd-cwt.md

                58(3): "80302"      / postcode ; REDACT /
            }
        }),
        58({                        / REDACT /
            500: true,              / inspection passed /
            502: 1612560720,        / 2021-02-04T20:14:00 /
            58(501): "EFGH-789012", / inspector license ; REDACT /
            58(503): {              / REDACT /
                1: "us",            / United States /
                58(2): "nv",        / region=Nevada ; REDACT /
                58(3): "89155"      / postcode ; REDACT /
            }
        }),
        58({                        / REDACT /
            500: true,              / inspection passed /
            502: 17183928,          / 2023-01-17T17:19:00 /
            58(501): "ABCD-123456", / inspector license ; REDACT /
            58(503): {              / REDACT /
                1: "us",            / United States /
                58(2): "ca",        / region=California ; REDACT /
                58(3): "94188"      / postcode ; REDACT /
            }
        })
    ]
}
~~~
{: #-nested-to-be-redacted title="EDN example of Nested Claims Set tagged with desired redactions"}

> The Issuer in our example respects the hints and produces the following Issued SD-CWT.

Author: ```suggestion
The Issuer in our example respects the hints and produces the following issued SD-CWT.
```

> It has generated 15 disclosures.

Author: ```suggestion
The Issuer has generated 15 disclosures.
```

> Note that the Issuer can list the issued disclosures (if any) in any order.

~~~ cbor-diag
{::include examples/nested_issuer_cwt.edn}
~~~
{: title="The Issued CWT containing nested redacted claims"}

Now the issued SD-CWT is provided to the Holder for processing.

## Holder

Once it receives an issued SD-CWT from the Issuer, the Holder validates all of the disclosures as described in {{holder-validation}}.

> The Holder needs to unwrap the nested claims starting with the top level and proceeding deeper, to insure that there are no unmatched Redacted Claim Hashes anywhere in the document, and no unmatched disclosures.

Author: ```suggestion
The Holder needs to unwrap all nested claims starting with the top level and proceeding deeper, to insure that there are no unmatched Redacted Claim Hashes anywhere in the document, and no unmatched disclosures.
```

Once the issued CWT is validated, the Holder can create multiple presentations, generating different SD-KBTs (as described in {{kbt}}) for each presentation by changing, for example, the audience and the choice of disclosures.
The privacy issues in {{privacy}} apply.

> In our example, the Holder chooses to disclose two of the inspections (from 2019 and 2023), the inspector license numbers and (partially redacted) location maps from both of these inspections, and the region (California) in the location map of the 2023 inspection.

Author: ```suggestion
In our example, the Holder chooses to disclose two of the inspections (from 2019 and 2023), the inspector license numbers and partially redacted location maps from both of these inspections, and the region (California) in the location map of the 2023 inspection.
```

In our example, the Holder sorts these disclosures in ascending order of the Redacted Claim Hash corresponding to each disclosure. (It could have used *any* order.)

~~~ cbor-diag
{::include examples/chosen-nested-disclosures.edn}
~~~
{: title="The Holder's choice of nested disclosures, sorted"}

Due to the way the claims are nested, disclosing the region of the 2023 inspection would not have been useful without disclosing the enclosing location map and its enclosing 2023 inspection array element.
In other words, the Holder needs to make sure that any intermediate levels of nested claims it wishes to disclose are also disclosed, otherwise the Verifier will not be able to validate them.

> At best, such "orphaned" disclosures will be discarded by the Verifier, and at worst, the Verifier could reject the entire SD-CWT.

Reviewer: What is the preferred strategy here?

The Holder then presents its SD-KBT to the Verifier.

## Verifier

The Verifier receives the Holder's SD-KBT, which contains an SD-CWT with the Holder's choice of (nested) disclosures.

> The example does not show the SD-KBT for brevity; that does not mean it would not be present.

Author: ```suggestion
The example does not show the SD-KBT for brevity; it is always used in an actual presentation.
```

In the example the disclosures are sorted in the order of the Redacted Claim Hash corresponding to each disclosure.
(The Holder can present its disclosures in any order.)

The Verifier needs to match the disclosures to their corresponding Redacted Claims in the Claims Set in the payload of our example SD-CWT.

> It needs to calculate the Redacted Claim Hash for each of the disclosures it receives.

Author: ```suggestion
The Verifier needs to calculate the Redacted Claim Hash for each of the disclosures it receives.
```

Typically the Verifier would create a lookup table of disclosures indexed by the Redacted Claim Hash.

> Comments in the example show the first 4 bytes of Redacted Claim Hash of each disclosure.

Author: ```suggestion
Comments in all the examples show the first 4 bytes of the Redacted Claim Hash of each disclosure.
```

The comments are only present in the diagnostic (text) notation to make it easier for the reader; comments are not included in CWTs or in SD-CWT in their native form.

Our example document only contains three Redacted Claims that are currently visible.

~~~ cbor-diag
...
/inspection history log/ 504: [
  / inspection 7-Feb-2019 /
  60(h'ca6b851688236744ff0cf0814508e4f1
       81d3811bfec4ed5bb8ace7823132dbc0'),
  / inspection 4-Feb-2021 /
  60(h'0dfdc69d0efb65eec14bf28af78cd8f0
       6cabfe8e2b1a0611ef8bca869ce35b61'),
  / inspection 17-Jan-2023 /
  60(h'20d9bb11363bae49851cfd4a3f166539
       d0aa00433c30aede18380bfa98d781dc')
]
...
~~~
{: title="Redacted Claims at Level 1"}

The Verifier looks for matching Redacted Claim Hashes among the disclosures it has received.

> In this example it finds two matching hashes (the first and last hash in the inspection history log claim array).

Reviewer: ```suggestion
In this example the Verifier finds two matching hashes (the first and last hash in the inspection history log claim array).
```

> The Verifier replaces the two matching Redacted Claim Hashes with their disclosed values (in this case, the last and first disclosures, which are both Claim Elements) and removes the second Redacted Claim Hash.

Author: ```suggestion
The Verifier replaces the two matching Redacted Claim Hashes with their disclosed values (in this case, the last and first disclosures, which are both replacing Redacted Claim Elements) and removes the second Redacted Claim Hash.
```

> The Verifier no longer considers the matched disclosures; they cannot match more than one Redacted Claim Hash.

Author: ```suggestion
The Verifier no longer considers the matched disclosures; each disclosure cannot match more than one Redacted Claim Hash.
```

~~~ cbor-diag
...
/inspection history log/ 504: [
  {
    500: true,
    502: 1549560720,
    simple(59): [
      h'7257a8697dfa40221079b00fb65fe587
        c310e6ca3da1aa33b090335de66ec810',
      h'c24c646b52fecd773c6ea01c6caa5a73
        422b85d3afa5900fa998336d83a88025'
    ]
  }   / inspection 7-Feb-2019 /,
  {
    500: true,
    502: 1674004740,
    simple(59): [
      h'af375dc3fba1d082448642c00be7b2f7
        bb05c9d8fb61cfc230ddfdfb4616a693',
      h'9d151abeb800adcc11ff10ff61fbd3d7
        5944c134b40a24abef1787d3ae6583aa'
    ]
  }   / inspection 17-Jan-2023 /
]
...
~~~
{: title="Claims after revealing disclosures at Level 1"}

The Verifier then repeats the process again at the next "level" of nesting.

> In this example all the visible Redacted Claims happen to be Redacted Claim Keys.

Reviewer: ```suggestion
In this example, all the visible Redacted Claims happen to be Redacted Claim Keys.
```

The second, third, fourth, and sixth disclosures match.
The Verifier replaces these Redacted Claims with the corresponding Claim Keys from the disclosures.

~~~ cbor-diag
...
/inspection history log/ 504: [
  {
    500: true,
    502: 1549560720,
    501: "DCBA-101777",
    503: {
      1: "us",
      simple(59): [
        h'1b89717167f39d51eec08b13baeda570
          eff5d0aedaa1d7d0821185c33634a5a0',
        h'49412884fa1e3787c17d1320bdd48f6e
          0e5365da010cde0571d4a7effd13cc2a'
      ]
    }
  }   / inspection 7-Feb-2019 /,
  {
    500: true,
    502: 1674004740,
    501: "ABCD-123456",
    503: {
      1: "us",
      simple(59): [
        h'2470fb9175b062c347ab3c3a19776d02
          476112a17cd7cfc9416664bc058c220b',
        h'cf397a08917528624ca3b332c9edcc54
          a72c9411dd5983f68017ce160f709f52'
      ]
    }
  }   / inspection 17-Jan-2023 /
]
...
~~~
{: title="Claims after revealing disclosures at Level 2"}

The Verifier then repeats the process again at the next "level" of nesting.
It finds a single matching disclosure (the second disclosure).

> It replaces the matching Redacted Claim and removes the others.

Author: ```suggestion
The Verifier finds a single matching disclosure (the second disclosure).
The Verifier replaces the matching Redacted Claim Hash and removes the others.
```

~~~ cbor-diag
...
/inspection history log/ 504: [
  {
    500: true,
    502: 1549560720,
    501: "DCBA-101777",
    503: {
      1: "us"
    }
  }   / inspection 7-Feb-2019 /,
  {
    500: true,
    502: 1674004740,
    501: "ABCD-123456",
    503: {
      1: "us",
      2: "ca" / region=California /
    }
  }   / inspection 17-Jan-2023 /
]
...
~~~
{: title="Claims after revealing disclosures at Level 3"}

The Verifier has no more Redacted Claim Hashes to process.

> Assuming the other validation steps pass, it can pass the Validated Claims Set on to the application.

Author: ```suggestion
Assuming the other validation steps pass, the Verifier can pass the Validated Claims Set on to the application.
```
